File: src/core/policy.ts
import { createHash, timingSafeEqual } from "node:crypto";
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { createRequire } from "node:module";
import type { AppConfig } from "./config.js";
import type { EngineType, PageSource } from "./models.js";

interface RobotsRobot {
  isAllowed(url: string, ua?: string): boolean | undefined;
}

const require = createRequire(import.meta.url);
const robotsParser = require("robots-parser") as (
  url: string,
  robotstxt: string,
) => RobotsRobot;

export function normalizeUrl(url: string): string {
  const u = new URL(url.trim());
  u.protocol = u.protocol.toLowerCase();
  u.hostname = u.hostname.toLowerCase();
  if (u.pathname !== "/" && u.pathname.endsWith("/")) {
    u.pathname = u.pathname.slice(0, -1);
  }
  return u.toString();
}

export function domainOf(url: string): string {
  return new URL(url).hostname.toLowerCase();
}

export function isSameDomain(url: string, base: string): boolean {
  return domainOf(url) === domainOf(base);
}

export function absoluteUrl(base: string, link: string): string | null {
  if (!link || link.startsWith("#") || /^(mailto|javascript|tel|data):/i.test(link)) return null;
  try {
    const joined = new URL(link, base).toString();
    const parsed = new URL(joined);
    if (!["http:", "https:"].includes(parsed.protocol)) return null;
    return normalizeUrl(joined);
  } catch {
    return null;
  }
}

export function contentHash(body: Buffer): string {
  return createHash("sha256").update(body).digest("hex");
}

export function saveContent(contentDir: string, url: string, body: Buffer): string {
  const digest = contentHash(body);
  const shard = digest.slice(0, 2);
  const dir = join(contentDir, shard);
  mkdirSync(dir, { recursive: true });
  const path = join(dir, `${digest}.html`);
  writeFileSync(path, body);
  return path;
}

export function extractTitle(html: string): string | null {
  const m = html.match(/<title[^>]*>([^<]+)<\/title>/i);
  return m ? m[1].trim() : null;
}

const JS_HINTS =
  /react|angular|vue|next\.js|nuxt|__NEXT_DATA__|window\.__INITIAL_STATE__/i;

export function needsJsRendering(html: string): boolean {
  if (html.length < 500) return true;
  const slice = html.slice(0, 50000);
  if (/<noscript/i.test(slice) && (slice.match(/<a\s+href=/gi) ?? []).length < 3) return true;
  return JS_HINTS.test(slice);
}

class RateLimiter {
  private interval: number;
  private lastRequest = new Map<string, number>();

  constructor(requestsPerSecond: number) {
    this.interval = 1000 / Math.max(requestsPerSecond, 0.1);
  }

  async acquire(host: string): Promise<void> {
    const now = Date.now();
    const last = this.lastRequest.get(host) ?? 0;
    const elapsed = now - last;
    if (elapsed < this.interval) {
      await sleep(this.interval - elapsed);
    }
    this.lastRequest.set(host, Date.now());
  }
}

class RobotsCache {
  private parsers = new Map<string, RobotsRobot | null>();
  private userAgent: string;
  private timeoutMs: number;

  constructor(userAgent: string, timeoutSeconds: number) {
    this.userAgent = userAgent;
    this.timeoutMs = timeoutSeconds * 1000;
  }

  async allowed(fetchFn: typeof fetch, url: string): Promise<boolean> {
    const host = domainOf(url);
    if (!this.parsers.has(host)) {
      this.parsers.set(host, await this.fetchParser(fetchFn, host));
    }
    const parser = this.parsers.get(host);
    if (!parser) return true;
    return parser.isAllowed(url, this.userAgent) ?? true;
  }

  private async fetchParser(fetchFn: typeof fetch, host: string) {
    const robotsUrl = `https://${host}/robots.txt`;
    try {
      const resp = await fetchFn(robotsUrl, {
        signal: AbortSignal.timeout(this.timeoutMs),
        headers: { "User-Agent": this.userAgent },
      });
      if (!resp.ok) return null;
      const text = await resp.text();
      return robotsParser(robotsUrl, text);
    } catch {
      return null;
    }
  }
}

export class CrawlPolicy {
  private config: AppConfig;
  private rateLimiter: RateLimiter;
  private robots: RobotsCache;

  constructor(config: AppConfig) {
    this.config = config;
    this.rateLimiter = new RateLimiter(config.policy.rateLimitPerHost);
    this.robots = new RobotsCache(
      config.orchestrator.userAgent,
      config.orchestrator.requestTimeoutSeconds,
    );
  }

  async preflight(
    fetchFn: typeof fetch,
    url: string,
    opts: { allowedDomains?: string[] | null; seedUrl: string },
  ): Promise<{ allowed: boolean; reason?: string }> {
    const parsed = new URL(url);
    if (!this.config.policy.allowedSchemes.includes(parsed.protocol.replace(":", ""))) {
      return { allowed: false, reason: "scheme_not_allowed" };
    }
    const host = parsed.hostname.toLowerCase();
    if (this.config.policy.blockedDomains.includes(host)) {
      return { allowed: false, reason: "domain_blocked" };
    }
    if (opts.allowedDomains?.length) {
      const ok = opts.allowedDomains.some(
        (d) => host === d || host.endsWith(`.${d}`),
      );
      if (!ok) {
        return { allowed: false, reason: "domain_not_allowed" };
      }
    }
    if (this.config.policy.respectRobotsTxt) {
      if (!(await this.robots.allowed(fetchFn, url))) {
        return { allowed: false, reason: "robots_disallowed" };
      }
    }
    await this.rateLimiter.acquire(host);
    return { allowed: true };
  }

  selectEngine(opts: {
    requested: EngineType;
    jsRendering: boolean;
    source: PageSource;
  }): EngineType {
    if (opts.requested !== "auto") return opts.requested;
    if (opts.source === "wayback" || opts.source === "common_crawl") return "archive";
    if (opts.jsRendering) return this.config.engines.routing.js_heavy as EngineType;
    return this.config.engines.default as EngineType;
  }
}

export function safeEqual(a: string, b: string): boolean {
  const ba = Buffer.from(a);
  const bb = Buffer.from(b);
  if (ba.length !== bb.length) return false;
  return timingSafeEqual(ba, bb);
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

File: omnispider/core/policy.py
from __future__ import annotations

import asyncio
import hashlib
import json
import re
import time
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from urllib.parse import urljoin, urlparse, urlunparse
from urllib.robotparser import RobotFileParser

import httpx
import structlog

from omnispider.core.config import AppConfig
from omnispider.core.models import EngineType, FetchResult, PageSource

log = structlog.get_logger()


def normalize_url(url: str) -> str:
    parsed = urlparse(url.strip())
    scheme = parsed.scheme.lower()
    netloc = parsed.netloc.lower()
    path = parsed.path or "/"
    if path != "/" and path.endswith("/"):
        path = path.rstrip("/")
    return urlunparse((scheme, netloc, path, "", parsed.query, ""))


def domain_of(url: str) -> str:
    return urlparse(url).netloc.lower()


def is_same_domain(url: str, base: str) -> bool:
    return domain_of(url) == domain_of(base)


class RateLimiter:
    def __init__(self, requests_per_second: float) -> None:
        self._interval = 1.0 / max(requests_per_second, 0.1)
        self._last_request: dict[str, float] = defaultdict(float)
        self._lock = asyncio.Lock()

    async def acquire(self, host: str) -> None:
        async with self._lock:
            now = time.monotonic()
            elapsed = now - self._last_request[host]
            if elapsed < self._interval:
                await asyncio.sleep(self._interval - elapsed)
            self._last_request[host] = time.monotonic()


class RobotsCache:
    def __init__(self, user_agent: str, timeout: float) -> None:
        self._user_agent = user_agent
        self._timeout = timeout
        self._parsers: dict[str, RobotFileParser | None] = {}
        self._lock = asyncio.Lock()

    async def allowed(self, client: httpx.AsyncClient, url: str) -> bool:
        host = domain_of(url)
        async with self._lock:
            if host not in self._parsers:
                self._parsers[host] = await self._fetch_parser(client, host)
            parser = self._parsers[host]
        if parser is None:
            return True
        return parser.can_fetch(self._user_agent, url)

    async def _fetch_parser(self, client: httpx.AsyncClient, host: str) -> RobotFileParser | None:
        robots_url = f"https://{host}/robots.txt"
        try:
            response = await client.get(robots_url, timeout=self._timeout, follow_redirects=True)
            if response.status_code >= 400:
                return None
            parser = RobotFileParser()
            parser.parse(response.text.splitlines())
            return parser
        except Exception as exc:
            log.debug("robots_fetch_failed", host=host, error=str(exc))
            return None


class CrawlPolicy:
    def __init__(self, config: AppConfig) -> None:
        self._config = config
        self._rate_limiter = RateLimiter(config.policy.rate_limit_per_host)
        self._robots = RobotsCache(
            config.orchestrator.user_agent,
            config.orchestrator.request_timeout_seconds,
        )

    async def preflight(
        self,
        client: httpx.AsyncClient,
        url: str,
        *,
        allowed_domains: list[str] | None,
        seed_url: str,
    ) -> tuple[bool, str | None]:
        parsed = urlparse(url)
        if parsed.scheme not in self._config.policy.allowed_schemes:
            return False, "scheme_not_allowed"
        host = parsed.netloc.lower()
        if host in self._config.policy.blocked_domains:
            return False, "domain_blocked"
        if allowed_domains and not any(host == d or host.endswith(f".{d}") for d in allowed_domains):
            if not self._config.policy.follow_external_links:
                return False, "external_domain"
            if allowed_domains and not is_same_domain(url, seed_url):
                return False, "domain_not_allowed"
        if self._config.policy.respect_robots_txt:
            if not await self._robots.allowed(client, url):
                return False, "robots_disallowed"
        await self._rate_limiter.acquire(host)
        return True, None

    def select_engine(
        self,
        *,
        requested: EngineType,
        js_rendering: bool,
        source: PageSource,
    ) -> EngineType:
        if requested != EngineType.AUTO:
            return requested
        if source in (PageSource.WAYBACK, PageSource.COMMON_CRAWL):
            return EngineType.ARCHIVE
        if js_rendering:
            return EngineType(self._config.engines.routing.js_heavy)
        return EngineType(self._config.engines.default)


_JS_HINTS = re.compile(
    r"(react|angular|vue|next\.js|nuxt|__NEXT_DATA__|window\.__INITIAL_STATE__)",
    re.IGNORECASE,
)


def needs_js_rendering(html: str) -> bool:
    if len(html) < 500:
        return True
    text = html[:50000]
    if "<noscript" in text.lower() and len(re.findall(r"<a\s+href=", text, re.I)) < 3:
        return True
    return bool(_JS_HINTS.search(text))


def content_hash(body: bytes) -> str:
    return hashlib.sha256(body).hexdigest()


def save_content(content_dir: Path, url: str, body: bytes) -> str:
    digest = content_hash(body)
    shard = digest[:2]
    target_dir = content_dir / shard
    target_dir.mkdir(parents=True, exist_ok=True)
    path = target_dir / f"{digest}.html"
    path.write_bytes(body)
    return str(path)


def extract_title(html: str) -> str | None:
    match = re.search(r"<title[^>]*>([^<]+)</title>", html, re.IGNORECASE)
    return match.group(1).strip() if match else None


def serialize_fetch_result(result: FetchResult) -> dict:
    return {
        "url": result.url,
        "final_url": result.final_url,
        "status_code": result.status_code,
        "content_type": result.content_type,
        "engine": result.engine.value,
        "source": result.source.value,
        "archive_timestamp": result.archive_timestamp,
        "error": result.error,
        "body_size": len(result.body),
    }


def load_json_lines(path: Path) -> list[dict]:
    if not path.exists():
        return []
    rows = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if line:
            rows.append(json.loads(line))
    return rows


def absolute_url(base: str, link: str) -> str | None:
    if not link or link.startswith(("#", "mailto:", "javascript:", "tel:", "data:")):
        return None
    joined = urljoin(base, link)
    parsed = urlparse(joined)
    if parsed.scheme not in ("http", "https"):
        return None
    return normalize_url(joined)
